
## Step 1:

Download accessions and keep in wkdir.

Create a directory to house all raw reads and download all raw reads to this directory.
Assuming sequencing data are short reads and paired end, split samples into foward and reverse reads for each accession. Make sure suffix ends in "_1.fastq" and "_2.fastq" for foward and reverse reads respectively. 

TO DO: 
allow different sequencing types
add more dynamic option to remove host
potentially add script to download and split sequencing (or add documention to literature)

## Step 2:

Create accession list and keep in wkdir.

## Step 3:

If mags are already prepared, keep in specified subdirectory in wkdir.

## Step 3:

Prepare a text file with paths to mag directories and keep in wkdir. 

## Step 4:

Run pipeline
